src/world_modeling/occupancy_segmentation/include/occupancy_segmentation_core.hpp

```diff
@@ -0,0 +1,377 @@
+#ifndef OCCUPANCY_SEGMENTATION_CORE_HPP_
+#define OCCUPANCY_SEGMENTATION_CORE_HPP_
+
+#include <pcl/common/centroid.h>
+#include <pcl/io/pcd_io.h>
+#include <pcl_conversions/pcl_conversions.h>
+#include <vector>
+
+#include <tbb/blocked_range.h>
+#include <tbb/parallel_for.h>
+#include <Eigen/Dense>
+
+#define NUM_ZONES 4
+#define L_MIN 2.7
+#define L_MAX 80.0
+
+#define Z_SEED 0.5
+#define MD 0.3
+#define MH -1.1
+
+#define MIN_NUM_POINTS 10
+#define NUM_SEED_POINTS 20
+#define TH_SEEDS 0.5
+#define UPRIGHTNESS_THRESH 0.5
+
+#define NUM_RINGS_OF_INTEREST 4
+#define SENSOR_HEIGHT 1.7
+#define GLOBAL_EL_THRESH 0
+
+struct PCAFeature {
+  Eigen::Vector3f principal_;
+  Eigen::Vector3f normal_;
+  Eigen::Vector3f singular_values_;
+  Eigen::Vector3f mean_;
+  float d_;
+  float th_dist_d_;
+  float linearity_;
+  float planarity_;
+};
+
+struct Patch_Index {
+  int idx;
+  int zone_idx;
+  int ring_idx;
+  int sector_idx;
+  int concentric_idx;
+};
+
+enum Status {
+  TOO_TILTED,
+  FLAT_ENOUGH,
+  TOO_HIGH_ELEV,
+  UPRIGHT_ENOUGH,
+  GLOBAL_TOO_HIGH_ELEV,
+  FEW_POINTS
+};
+
+template <typename PointT>
+class OccupancySegmentationCore {
+ public:
+  typedef std::vector<pcl::PointCloud<PointT>> Ring;
+  typedef std::vector<Ring> Zone;
+  // Size of buffer before processed messages are published.
+  static constexpr int BUFFER_CAPACITY = 10;
+
+  const int ZONE_RINGS[NUM_ZONES] = {2, 4, 4, 4};
+  const int ZONE_SECTORS[NUM_ZONES] = {16, 32, 54, 32};
+  const float FLATNESS_THR[NUM_ZONES] = {0.0005, 0.000725, 0.001, 0.001};
+  const float ELEVATION_THR[NUM_ZONES] = {0.523, 0.746, 0.879, 1.125};
+  const double lmins[NUM_ZONES] = {L_MIN, (7 * L_MIN + L_MAX) / 8, (3 * L_MIN + L_MAX) / 4,
+                                   (L_MIN + L_MAX) / 2};
+  const double lmaxs[NUM_ZONES] = {lmins[1], lmins[2], lmins[3], L_MAX};
```

Comment:
seems like a bunch of magic numbers? correct me if these are rarely changed

Reply:
lmins and lmaxs won't be changed, but the others can be changed as hyperparameters for the framework. Do you want these also ROS parameterized?

Reply:
done